address issue #854 : don't allow mv/cp of multiple files to single file

guygurari opened the pull request:
When trying to cp/mv multiple files, make sure the destination
has a trailing slash so it is treated as a directory. Otherwise,
multiple files can get moved to a single file, leading to data
loss.

fviard commented:
Thank you for your contribution.
Regarding the fix for the tests, that is good, but that would be great if you can also backport your changes to run-tests-minio.py. For the moment, it is a little duplicates, but it is the "minio" one that is run as "automatic" tests by travisci.

Regarding the fix itself, i did think more about this and in my opinion, the best would be to throw something like: (remote2local)
raise ParameterError("Destination must be a directory and end with '/' when downloading multiple sources.")
When you encounter the bad case. So things are explicit and not implicit.
And with that, maybe you can get ride of the "cfg.recursive" and have a single behavior for all the case.
in1 out -> ok
in1 out/ -> ok
in1/ out -> wrong, raise error
in1 in2 out -> wrong, raise error
in1 in2 out/ -> ok
in1/ out/ -> ok

People currently using multiple sources with recursive and no "/" for destination could encounter a behavior change. But it would be an explicit error and the fix for them would be to add a "/" at destination, so I think it should be ok.

guygurari commented:
Thanks for looking at this. I changed the behavior to raise an exception when trying to move 'in1 in2 -> out' or when moving recursively to 'out' (without a '/'). I also added the tests to `run-tests-minio.py`.

I did not see a way to ignore `cfg.recursive`, because moving `in -> out` should be allowed when it moving non-recursively. Also, the case 'in1/ -> out' is already handled in `S3.FileLists` before it reaches this code, so I did not handle it.

fviard commented:
I'm sorry to take a little time to reply to you, i have a little back log that accumulated.
Thank you for the update.
I will review that in the coming days.

guygurari commented:
No worries, I know we're all busy :-)

fviard commented:
Merged! Thank you very much for your contribution, to have taken care of adding auto tests and for your patience!

guygurari commented:
Great! Glad I could contribute.
